[MRG] FIX TFIDFVectorizer follow parameter dtype and TFIDFTransformer preserve dtype

maykulkarni opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
Fixes #10411 #8044
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.
- [x] TfidfVectorizer now respects data type
- [x] TfidfVectorizer validates intput and casts if int*
- [x] 2 test cases to check the changes

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

maykulkarni committed:
Fixes #10310: Cross Validation FitFailedWarning message and exception formatting

maykulkarni committed:
Fixed a doctest where the error message was different

maykulkarni committed:
Fixed doctest to pass test

maykulkarni committed:
Fixed more test

maykulkarni committed:
Fixed some more errors in tests

maykulkarni committed:
typo in last commit

maykulkarni committed:
Added imports globally

maykulkarni committed:
Fixes #10411 TfidfVectorizer doesn't repect dtype

maykulkarni committed:
messed up last commit

maykulkarni committed:
mend

maykulkarni committed:
Merge branch 'master' of https://github.com/maykulkarni/scikit-learn

maykulkarni committed:
messed up last commit

maykulkarni committed:
Removed a different PR

maykulkarni committed:
Fixes #10411 TfidfVectorizer doesn't repect data type

maykulkarni committed:
Changes .rst file doc tests

glemaitre commented on sklearn/feature_extraction/text.py:
```diff
@@ -1078,7 +1078,7 @@ def fit(self, X, y=None):
             # log+1 instead of log makes sure terms with zero idf don't get
             # suppressed entirely.
-            idf = np.log(float(n_samples) / df) + 1.0
+            idf = np.log(float(n_samples) / df).astype(X.dtype) + 1.0
```
I don't like this line:

`np.log(n_samples / df, dtype=X.dtype) + 1` seems less clutter to me. I don't know if df can be a scalar which in this case you might want to import division from future in this case

rth replied:
Also X can be an `np.int*` in which case you should fall back float64 for backward compatibility, I think.

maykulkarni replied:
Yes exactly! using `np.log(n_samples / df, dtype=X.dtype)` will raise a TypeError in case the datatype is int*. Using `np.log(float(n_samples) / df).astype(X.dtype)` will automatically fallback to np.float64 in case the conversion isn't possible

maykulkarni replied:
So should I leave it that way or use if else to check if it's int* and use the definition given @glemaitre

glemaitre reviewed (changes requested):
A first review

rth commented on doc/modules/feature_extraction.rst:
```diff
@@ -463,12 +463,12 @@ content of the documents::
-         [ 0.47330339,  0.88089948,  0.        ],
-         [ 0.58149261,  0.        ,  0.81355169]])
+         [ 0.6       ,  0.8       ,  0.        ],
+         [ 0.6       ,  0.        ,  0.8       ]])
```
The fact that the value of the computed array changed in examples means that backward compatibility is broken. This shouldn't happen.

maykulkarni committed:
Made changes according to review

glemaitre commented on sklearn/feature_extraction/tests/test_text.py:
```diff
@@ -995,3 +998,26 @@ def test_vectorizer_string_object_as_input():
+    assert X.dtype == X_trans.dtype
+
+    # another case with string
+    test = TfidfVectorizer(dtype=dtype)
```
test -> vectorizer

glemaitre commented on sklearn/feature_extraction/text.py:
```diff
@@ -1078,7 +1078,8 @@ def fit(self, X, y=None):
             # suppressed entirely.
-            idf = np.log(float(n_samples) / df) + 1.0
+            # astype will fallback to np.float64 if X.dtype is int
+            idf = np.log(float(n_samples) / df).astype(X.dtype) + 1.0
```
We don't preserve backward compatibility if X is int (it is true that we keep the type thought).

```python
In [6]: np.log(5. / np.array([1, 2, 3.5])).astype(np.int32) + 1.0
Out[6]: array([ 2.,  1.,  1.])

In [7]: np.log(5. / np.array([1, 2, 3.5])) + 1.0
Out[7]: array([ 2.60943791,  1.91629073,  1.35667494])
```

The results are really different due to the casting.

glemaitre replied:
By the way I am not sure that `float(n_samples)` is useful `n_samples` should be enough.

@rth Could you confirm that my thoughts are right (I don't know the code base of the TD-IDF) in fact.

rth commented:
> The results are really different due to the casting.

@maykulkarni Because the current approach doesn't produce correct results, I think it might be better to just cast the input array to float32/64 at validation time, see e.g. https://github.com/scikit-learn/scikit-learn/blob/b94876a3a20e1d629698ea1c7b358169b97bfcce/sklearn/preprocessing/data.py#L621-L622 This would also allow addressing https://github.com/scikit-learn/scikit-learn/issues/8044

> By the way I am not sure that float(n_samples) is useful n_samples should be enough.

@glemaitre My interpretation, is that that was used to make `float(n_samples) / df` a float division in case `df.dtype == int`. If we cast `X` to float beforehand, this should be no longer necessary, think..

rth commented:
Though maybe the performance of https://github.com/scikit-learn/scikit-learn/blob/b94876a3a20e1d629698ea1c7b358169b97bfcce/sklearn/feature_extraction/text.py#L1073 would be affected if X was cast to float at the beginning of the `fit`. By default, CountVectorizer produces int64..

maykulkarni committed:
Completed the changes requested

Signed-off-by: Mayur Kulkarni <mayurkulkarni012@gmail.com>

maykulkarni commented:
I've added the changes suggested by @rth. While the backward compatibility is still broken, I've managed to preserve some of the results that were changed in the last commit.

maykulkarni committed:
Added changes requested

maykulkarni commented:
The test passes for Python 2 but fails for Python 3. If I changed the cases to comply, the test fails for Python 2 and passes for Python 3. Am I missing something?

rth commented:
I think, the desired behaviour, if for,
*  float64, int32, int64 to  keep the same behaviour as before for backward compatibility
* float32 to return float32 which corresponds to this bug fix and that won't be backward compatible. 

So any tests or or doctest that fail, it's necessary to check the dtype of the input. If it was float32 it's OK to change the expected results. If it wasn't we have to make them pass as they are. In particular, I think you might have to  revert the doctests to their original version.

BTW, to avoid re-running the whole test suite each time locally, you can either provide `pytest` with the paths to the relevant files or use the `-k` selector flag..

rth commented on sklearn/feature_extraction/text.py:
```diff
@@ -1078,7 +1083,8 @@ def fit(self, X, y=None):
             # suppressed entirely.
-            idf = np.log(float(n_samples) / df) + 1.0
+            # astype will fallback to np.float64 if X.dtype is np.int*
+            idf = np.log(n_samples / df).astype(X.dtype) + 1.0
```
The `astype` is no longer necessary IMO.

maykulkarni committed:
reset doctests to ensure backward compatibility

maykulkarni commented:
After many hours of pondering, I realized that the culprit was `_document_frequency(X)` which used to return int32 that breaks the tests. That was the reason why the tests failed  after converting `idf` to X.dtype. But strangely enough, the tests were passing for Python 2 and failing for Python 3.x. Casting it to `X.dtype` not only passed the tests but also ensured backward compatibility.

glemaitre commented on sklearn/feature_extraction/text.py:
```diff
@@ -1068,17 +1068,23 @@ def fit(self, X, y=None):
             X = sp.csc_matrix(X)
+
+        # force convert to np.float*
+        X = check_array(X, accept_sparse='csc', warn_on_dtype=True,
```
This will convert X to a float array. Do we really want that.

maykulkarni replied:
Yes, this was suggested by @rth. Also, if we don't, then the test cases don't comply with int* values. Furthermore, see #8044

jnothman replied:
I've not reviewed the discussion, but I don't get why we are casting X in fit. Can't we just cast df?

rth replied:
It's discussed in https://github.com/scikit-learn/scikit-learn/pull/10443#issuecomment-357459029 and in the 3  following comments.

Your opinion on that would we helpful.

jnothman replied:
I still don't get why TfidfTransformer.fit needs to copy X unless X is not CSR/CSC. It should use the default dtype='numeric' in check_array, then use `df = df.astype(X.dtype if X.dtype.kind == 'f' else np.float64)`.

glemaitre replied:
> dtype='numeric'

yep, I messed up [there](https://github.com/glemaitre/scikit-learn/commit/4e5e1f06ed1b90c0ffb00584db81a4e8c77e1dff#diff-904f5f849e36c648b8b05de140d7fad8R1076) but it was what I intended with `DTYPES`

glemaitre replied:
> df = df.astype(X.dtype if X.dtype.kind == 'f' else np.float64)

we could avoid that by using dtype in log

```python
np.log(n_samples / df, dtype=np.float64) + 1
```

glemaitre commented:
@rth I getting a bit confused here. Why should we trigger a copy of X with the check_array. I would think that we want to avoid such thing. Also, one thing which is not trivial to me is the `dtype` argument. Why is it not passed from the tf-idf vectorizer to the transformer.

My thoughts were to avoid useless copy and ensure that `transform` and `fit_transform` would return what they are supposed to return, isn't it.

rth commented:
> I getting a bit confused here. Why should we trigger a copy of X with the check_array. I would think that we want to avoid such thing. Also, one thing which is not trivial to me is the dtype argument. Why is it not passed from the tf-idf vectorizer to the transformer?

If we set performance aside for a second, from the API consistency point of view, I think, it would make sense for `TfidfTransformer` to behave like `StandardScaler` or `MinMaxScaler` since they are all scaling features in some way. This would mean no `dype` init parameter; dtype of output ==  dtype of input if possible, float otherwise.  

`TfidfVectorizer` is basically a pipeline of `CountVectorizer` with a `TfidfTransformer`. If we want float64 as output, the logical thing would be to for the `CountVectorizer` there to be initialized with `dtype=float64` which will build a sparse float64 [matrix without too much memory copies](https://github.com/scikit-learn/scikit-learn/blob/a24c8b464d094d2c468a16ea9f8bf8d42d949f84/sklearn/feature_extraction/text.py#L820), and `check_array` in  `TfidfTransformer` also won't trigger a copy. This is achievable by having a default init value `TfidfVectorizer(...,  dtype=np.float64)` as it's the case e.g. in the `HashingVectorizer`, the current default value of `dtype=np.int64` does not make sense and is a bug IMO.  We can change this default without a deprecation cycle since the dtype parameter was not used anyway, right?

In the case when we explicitly build a pipeline with `make_pipeline(CountVectorizer(), TfidfTransformer())`, this approach will be indeed less efficient, because, as you mentioned the `check_array` would trigger a conversion from int to float64.  Several possibilities here,
 1. do nothing
 2. Raise a performance warning that the vectorizer earlier in the pipeline, probably  `CountVectorizer()` should be initialized with `dtype=np.float`
 3. In `TfidfTranformer`, extend the `dtype` argument in `check_array` to accept `int32/64` and implement some more custom logic to handle those. 

I agree that 3. might be useful. What do you think?

glemaitre commented:
Ok, so this is not so much trivial.

I coded what I would have in mind: https://github.com/glemaitre/scikit-learn/commit/96d419f51ece367b3f77dcfecf52931605c90577

My idea would be to avoid all copy/conversion in the `TfidfTransformer` and always return np.float32 or np.float64 dtype after transform

Since X can be an Integer type and that the output will be a float anyway, I think that the Transformer should required a `dtype` argument. I know that you mentioned that usually Transformers do not have any of those arguments but there is some example, e.g. CategoricalEncoder.

By adding this keyword, it allows to modify the `TfidfVectorizer` by forcing the type of the Transformer instead of the Vectorizer. It does not seem that I am breaking the backward compatibility

@rth Do you see any issue with this approach?

glemaitre commented:
The output of Travis should be there: https://travis-ci.org/glemaitre/scikit-learn/builds/329115473

rth commented:
> Ok, so this is not so much trivial.

I also thought that this would be more straightforward. Sorry about that @maykulkarni .

> Since X can be an Integer type and that the output will be a float anyway, I think that the Transformer should required a dtype argument. I know that you mentioned that usually Transformers do not have any of those arguments but there is some example, e.g. CategoricalEncoder.

Fair enough, that would also allow having more degrees of freedom for dtypes between the vectorizer and the transformer which is probably a good thing (requires fewer thoughts as how to get the desired output).  

Maybe @jnothman would have an opinion on this as well? Just to make sure there is a consensus.

> Do you see any issue with this approach?

I won't be able to review in detail before a few days, but generally, if tests pass that seem promising. There is a single doctest failure in `doc/modules/feature_extraction.rst:525: DocTestFailure` It might be also worth running the `bench_20newsgroups.py` to see if this affects anything.

maykulkarni commented:
The failed test indicates broken backward compatibility right? I was able to achieve the same results before forcing the input to convert to float

jnothman commented:
Which failed test??

jnothman reviewed (commented):
You're saying that before we always returned float64 results and hence changing df to float32 breaks backwards compatibility? Yes, that seems right. I suspect it would be less of an issue if we applied the cast to `_idf_diag` rather than `df`

maykulkarni commented:
@jnothman 
> Which failed test??

The build by @glemaitre where he returned np.float32 or np.float64 after transform

> I suspect it would be less of an issue if we applied the cast to _idf_diag rather than df

Seems good to me too, what are your thoughts @glemaitre @rth ?

jnothman commented:
I suggest you try it and check if it passes tests. I presume the loss in
precision in the failing test was due to taking the log in low precision.

glemaitre commented:
> I suggest you try it and check if it passes tests. I presume the loss in precision in the failing test was due to taking the log in low precision.

Yep it could be the case since I removed `float` in the log

glemaitre commented:
> I suggest you try it and check if it passes tests. I presume the loss in precision in the failing test was due to taking the log in low precision.

maykulkarni committed:
Made changes according to review

maykulkarni commented:
It turns out @jnothman was right, taking log in lower precision was causing the problem. I've removed the forcible conversion to np.float and replaced it with a cast iff input type is non float.

glemaitre commented on sklearn/feature_extraction/tests/test_text.py:
```diff
@@ -995,3 +998,28 @@ def test_vectorizer_string_object_as_input():
+
+
+@pytest.mark.parametrize("dtype", [
+    np.float32,
```
you need to parametrize the dtype of X and the dtype of the tfidfvectorizer to try all combination.

I would suspect that right now:

* X -> np.float32, Vectorizer -> np.float64
* X -> np.float64, Vectorizer -> np.float32

will failed.

maykulkarni replied:
I've added separate cases for Vectorizer and Transformer. I hope you mean the same thing?

maykulkarni committed:
Added separate test for vectorizer and transformer

maykulkarni commented:
@glemaitre @jnothman review please

jnothman commented on sklearn/feature_extraction/text.py:
```diff
@@ -1068,17 +1068,21 @@ def fit(self, X, y=None):
             # log+1 instead of log makes sure terms with zero idf don't get
             # suppressed entirely.
-            idf = np.log(float(n_samples) / df) + 1.0
+            idf = np.log(n_samples / df) + 1.0
```
I would have thought we should always perform this log in float64, and only afterwards convert back to float32 if appropriate. Confused. I don't understand why you had that test failing before and not now.

maykulkarni replied:
Because `df` is converted to float if it's in int

jnothman replied:
But shouldn't it be converted to float64 in all cases, then idf converted back to float32 if necessary?

maykulkarni replied:
Yeah it seems a little magical that the dtype after transforming matches even if I am not converting it back to the original type. I'll do a little debugging to answer this question.

jnothman replied:
But you're also unnecessarily reducing the precision of the calculation here in the float32 case.

maykulkarni replied:
OK so the required behavior is always convert `df` to float64 for precision and convert `idf` back to X.dtype?

glemaitre replied:
It was my proposal in https://github.com/glemaitre/scikit-learn/commit/4e5e1f06ed1b90c0ffb00584db81a4e8c77e1dff

maykulkarni commented:
@jnothman Sorry, I mean convert is back to X.dtype if it is float32 else leave it float64

jnothman commented:
Yes

glemaitre commented on sklearn/feature_extraction/tests/test_text.py:
```diff
@@ -995,3 +998,34 @@ def test_vectorizer_string_object_as_input():
+    np.float64
+])
+def test_tfidf_transformer_type(dtype):
+    # create made up data with parameterized dtype
```
remove this comment

glemaitre commented on sklearn/feature_extraction/tests/test_text.py:
```diff
@@ -995,3 +998,34 @@ def test_vectorizer_string_object_as_input():
+    # create made up data with parameterized dtype
+    X = rand(10, 20000, dtype=dtype, random_state=42)
+    X_trans = TfidfTransformer().fit_transform(X)
+    # check if the required data types match
```
remove this comment

glemaitre commented on sklearn/feature_extraction/tests/test_text.py:
```diff
@@ -995,3 +998,34 @@ def test_vectorizer_string_object_as_input():
+    assert dtype == X_idf.dtype
+
+
+def test_tfidf_fallback():
```
So you probably need to create  X in the parametrize and to pass the dtype as well.

glemaitre commented on sklearn/feature_extraction/tests/test_text.py:
```diff
@@ -995,3 +998,34 @@ def test_vectorizer_string_object_as_input():
+    X = rand(10, 20000, dtype=dtype, random_state=42)
+    X_trans = TfidfTransformer().fit_transform(X)
+    # check if the required data types match
+    assert X.dtype == X_trans.dtype
```
Usually we write `x == expected`

```python
assert X_trans.dtype == X.dtype
```

glemaitre commented on sklearn/feature_extraction/tests/test_text.py:
```diff
@@ -995,3 +998,34 @@ def test_vectorizer_string_object_as_input():
+@pytest.mark.parametrize("dtype", [
+    np.float32,
+    np.float64,
+])
```
Try all combination which are possible:

```python
@pytest.mark.parametrize(
    "vectorizer_dtype, output_dtype",
    [(np.int32, np.float64),
     (np.int64, np.float64),
     (np.float32, np.float32),
     (np.float64, np.float64)])
````

glemaitre commented on sklearn/feature_extraction/tests/test_text.py:
```diff
@@ -995,3 +998,34 @@ def test_vectorizer_string_object_as_input():
+])
+def test_tfidf_vectorizer_type(dtype):
+    X = np.array(["numpy", "scipy", "sklearn"])
+    vectorizer = TfidfVectorizer(dtype=dtype)
```
pass `vectorizer_dtype`

glemaitre commented on sklearn/feature_extraction/tests/test_text.py:
```diff
@@ -995,3 +998,34 @@ def test_vectorizer_string_object_as_input():
+    X = np.array(["numpy", "scipy", "sklearn"])
+    vectorizer = TfidfVectorizer(dtype=dtype)
+    X_idf = vectorizer.fit_transform(X)
+    assert dtype == X_idf.dtype
```
inverse the checking

```python
assert X_idf.dtype = output_dtype
```

glemaitre commented on sklearn/feature_extraction/text.py:
```diff
@@ -1068,17 +1068,21 @@ def fit(self, X, y=None):
             # suppressed entirely.
-            idf = np.log(float(n_samples) / df) + 1.0
+            idf = np.log(n_samples / df) + 1.0
             self._idf_diag = sp.spdiags(idf, diags=0, m=n_features,
```
As mentioned by @jnothman 

```python
self._idf_diag = sp.diags(idf, offsets=0,
                                      shape=(n_features, n_features),
                                      format='csr',
                                      dtype=X.dtype
                                      if X.dtype in FLOAT_DTYPES
                                      else np.float64)
```

glemaitre commented:
In addition, you have to add some `check_array` at the beginning of `fit`:
```python
X = check_array(X, accept_sparse='csc', dtype='numeric')
```
and `transform`:
```python
X = check_array(X, accept_sparse='csr', dtype=FLOAT_DTYPES)
if not sp.issparse(X):
    X =  sp.csr_matrix(X, dtype=np.float64, copy=copy)
```
You can also avoid a call to the sparse constructor in `document_frequency`.
```python
def _document_frequency(X):
     """Count the number of non-zero values for each feature in sparse X."""
     if sp.isspmatrix_csr(X):
         return np.bincount(X.indices, minlength=X.shape[1])
     else:
         return np.diff(X.indptr)
```

 The previous issue in log was due to the division because `from __future__ import division` was not done.

glemaitre commented:
@maykulkarni do you think that you can fix this PR?

maykulkarni commented:
Hey hi, yes. I was stuck in some office work. I'll get in done tomorrow.

jnothman commented:
Should we find someone to help with this @maykulkarni ?

rth commented:
No worries, thanks for your contribution @maykulkarni ! 

I think @glemaitre mostly arrived at a working solution while reviewing this PR and might be willing to continue this one..

glemaitre commented:
I am on it ;)

On 16 May 2018 at 12:04, Roman Yurchak <notifications@github.com> wrote:

> No worries, thanks for your contribution @maykulkarni
> <https://github.com/maykulkarni> !
>
> I think @glemaitre <https://github.com/glemaitre> mostly arrived at a
> working solution while reviewing this PR and might be willing to continue
> this one..
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/10443#issuecomment-389465558>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHG9P-tCoRjogueVtNfvtUfREUJUgS1Oks5ty_mfgaJpZM4RYyAs>
> .
>



-- 
Guillaume Lemaitre
INRIA Saclay - Parietal team
Center for Data Science Paris-Saclay
https://glemaitre.github.io/

glemaitre committed:
fixup

glemaitre committed:
Merge remote-tracking branch 'origin/master' into maykulkarni-TfidfVectorizerFix

glemaitre committed:
DOC add whats new

glemaitre commented:
@rth @jnothman I am approving my own changes but it would be better that you review this PR instead :)

rth commented on sklearn/feature_extraction/text.py:
```diff
@@ -1117,21 +1117,30 @@ def fit(self, X, y=None):
         X : sparse matrix, [n_samples, n_features]
             a matrix of term/token counts
         """
+        X = check_array(X, accept_sparse='csc')
```
Why casting to "csc" in fit ? Typical input would be "csr", wouldn't it?

glemaitre replied:
The only place that `X` is used is in `_document_frequency` which expect a `csc` matrix.

glemaitre replied:
FYI: https://github.com/scikit-learn/scikit-learn/pull/10443/files#diff-904f5f849e36c648b8b05de140d7fad8L576

rth replied:
Maybe I am missing something but `_document_frequency` works fine with CSR matrices https://github.com/scikit-learn/scikit-learn/blob/48f3303bfc0be26136b98e9aa95dc3b3f916daff/sklearn/feature_extraction/text.py#L571-L574

glemaitre replied:
Uhm good point

rth commented:
Thanks @glemaitre ! Will review later today.

jnothman commented on sklearn/feature_extraction/tests/test_text.py:
```diff
@@ -1042,6 +1045,27 @@ def test_vectorizer_string_object_as_input():
+
+@pytest.mark.parametrize(
+    "vectorizer_dtype, output_dtype",
+    [(np.int32, np.float64),
```
What's the sense in allowing an integer dtype to tfidf?

glemaitre replied:
This is the `dtype` of the vectorizer and not of the transformer. So it can be integer even if the transformer will return float.

rth replied:
Yes, so essentially when `dtype` is not either float32 or float64 this parameter has no effect. Shouldn't a warning or error be raised in that case then (rather than silently ignoring a user-provided parameter)?

glemaitre replied:
The funny part here is that the default value of `dtype` is `np.int64`. It seems wrong then.
Shall we make it `np.float64`

rth commented on sklearn/feature_extraction/text.py:
```diff
@@ -1117,21 +1117,30 @@ def fit(self, X, y=None):
+                                      format='csr',
+                                      dtype=X.dtype
+                                      if X.dtype in FLOAT_DTYPES
+                                      else np.float64)
```
Maybe factorize this as `df_dtype = X.dtype if ...` as it's used twice and is a bit verbose

jnothman commented:
yes i think we should be warning for or rejecting integer dtype. The dtype
for CountVectorizer is not relevant and indeed it should always be using
the one that is most efficient for tfidf transformation

glemaitre committed:
address rth and jnothman comments

glemaitre commented:
In the last commit, I did:

* raise a warning if passing integer dtype.
* change the default dtype to float64 since it should not change anything.
* change the conversion from CSC to CSR at fit time when a dense matrix is given. Basically, up to now it seems that we had CSC constructor just for historical reason (_document_frequency was expecting such matrix which was then changed for both type but not the construction).
* change the test to check that we raise the warning.

@rth @jnothman does it make sense.

jnothman commented on sklearn/feature_extraction/text.py:
```diff
@@ -1117,21 +1118,26 @@ def fit(self, X, y=None):
         if not sp.issparse(X):
-            X = sp.csc_matrix(X)
+            X = sp.csr_matrix(X)
+        dtype = X.dtype if X.dtype in FLOAT_DTYPES else np.float64
```
For CountVectorizer it still makes sense to support integer dtype

glemaitre replied:
Ok but this is the TfidfTransformer and we will output float64 here.

jnothman replied:
So it is!

glemaitre committed:
spelling

glemaitre committed:
simplify tests

jnothman reviewed (approved):
LGTM

glemaitre commented:
@rth Could you have a look.

rth commented on doc/whats_new/v0.20.rst:
```diff
@@ -448,6 +448,12 @@ Feature Extraction
 
+- Fixed bug in :class:`feature_extraction.text.TFIDFVectorizer` in which the
+  output of the vectorizer will be defined by the parameter ``dtype``. In
+  addition, :class:`feature_extraction.text.TFIDFTransformer` will preserver
```
maybe "now preserves"

rth commented on doc/whats_new/v0.20.rst:
```diff
@@ -448,6 +448,12 @@ Feature Extraction
   and `Roman Yurchak`_.
 
+- Fixed bug in :class:`feature_extraction.text.TFIDFVectorizer` in which the
+  output of the vectorizer will be defined by the parameter ``dtype``. In
```
Maybe,
```rst
Fixed bug in :class:`feature_extraction.text.TFIDFVectorizer` which 
was ignoring the parameter ``dtype``. 
```
?

rth commented on sklearn/feature_extraction/text.py:
```diff
@@ -1151,11 +1157,8 @@ def transform(self, X, copy=True):
-            X = sp.csr_matrix(X, copy=copy)
-        else:
-            # convert counts or binary occurrences to floats
+        X = check_array(X, accept_sparse='csr', dtype=FLOAT_DTYPES)
```
add `copy=copy` otherwise this will ignore the copy parameter passed to transform I think?

glemaitre replied:
right and I can remove from the `if sparse...`

rth commented on sklearn/feature_extraction/tests/test_text.py:
```diff
@@ -1042,6 +1045,29 @@ def test_vectorizer_string_object_as_input():
+    X = np.array(["numpy", "scipy", "sklearn"])
+    vectorizer = TfidfVectorizer(dtype=vectorizer_dtype)
+    with pytest.warns(expected_warning, match=msg_warning):
+            X_idf = vectorizer.fit_transform(X)
```
To check that no warning is raised (if that's what you want to do) you could use https://stackoverflow.com/a/45671804/1791279. I wonder if,
```py
    with pytest.warns(expected_warning, match=msg_warning) as record:
            X_idf = vectorizer.fit_transform(X)
    if expected_warning is None:
        assert len(record) == 0
```
would work.

glemaitre replied:
Right, I was expecting that it works but actually we would need this: https://github.com/pytest-dev/pytest/issues/1830#issuecomment-378357212 to be implemented.

rth commented on sklearn/feature_extraction/text.py:
```diff
@@ -573,7 +574,7 @@ def _document_frequency(X):
         return np.bincount(X.indices, minlength=X.shape[1])
     else:
-        return np.diff(sp.csc_matrix(X, copy=False).indptr)
+        return np.diff(X.indptr)
```
Checked that here we only end-up here with CSC matrices, so the `csc_matrix` is indeed redundant.

Still it's not obvious when reading the code. I would either add a comment, or keep it as it was (I guess it will simply validate sparse CSC array type with `copy=False`).

**Edit:**  this line is not covered, so if we do change it, we would need to add a test.

rth replied:
After more thoughts, it might make sense to keep it as it was.

Otherwise, if someone were to provide `scipy.sparse.bsr_matrix` to it (e.g. if the code base changes in the future or someone is using this private method) it will silently produce incorrect results, I think.

There is no real advantage in removing it, short of reduced verbosity, right?

glemaitre replied:
That might be slower:

```python
In [5]: X = sparse.random(100000, 1000).tocsc()

In [7]: %timeit np.diff(X.indptr)
3.65 µs ± 42.7 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [8]: %timeit np.diff(sparse.csc_matrix(X, copy=False).indptr)
33.6 µs ± 3.27 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

```

glemaitre replied:
bsr_matrix will be converted by the `check_array`

rth replied:
> bsr_matrix will be converted by the check_array

I know, it was more a general comment of not changing something if there is no need to. 

Fair enough, the speedup is significant.

rth reviewed (commented):
A few more comments below, otherwise this LGTM.

glemaitre committed:
rth comments

glemaitre committed:
TST check consistency between csc and csr matrix

glemaitre commented:
@rth I am just unsure about the `X.indptr`. I don't know if a x10 difference there would be an issue.

rth commented on doc/whats_new/v0.20.rst:
```diff
@@ -448,6 +448,11 @@ Feature Extraction
 
+- Fixed bug in :class:`feature_extraction.text.TFIDFVectorizer` which 
+  was ignoring the parameter ``dtype``. 
+  :issue:`10441` by :user:`Mayur Kulkarni <maykulkarni>` and
```
You don't want to mention `TFIDFTransformer` anymore?

glemaitre replied:
ups I messed up.

rth commented on sklearn/feature_extraction/tests/test_text.py:
```diff
@@ -1,6 +1,10 @@
 
+import pytest
+from scipy import sparse
+from scipy.sparse import rand
```
maybe just use `sparse.rand` then ?

rth commented:
> I am just unsure about the X.indptr. I don't know if a x10 difference there would be an issue.

You are right, let's keep the proposed changes.

I added a few more comments regarding the newer commits.

glemaitre committed:
FIX roman comments

rth commented:
Thanks @glemaitre ! LTGM, will merge when the CI is green.

jnothman commented:
Thanks @glemaitre 

When I merged, I noticed you used "Fixes #10411 #8044" in the description. GitHub is not that clever. Use "Fixes #10411. Fixes #8044."

david-waterworth commented:
This still seems to be an issue? When I pass `dtype=np.float32` to the constructor of  `TfidfVectorizer` then call tf.fit_transform(docs) the result is a "sparse matrix of type '<class 'numpy.float64'>'"

XinY-Z commented:
Same as @david-waterworth, when I tried to use `dtype==np.float32` for the `TfidfVectorizer`,  it gives a warning 
`UserWarning: Only (<class 'numpy.float64'>, <class 'numpy.float32'>, <class 'numpy.float16'>) 'dtype' should be used. numpy.float32 'dtype' will be converted to np.float64.`.
